File: rust/crates/sera-core/src/routes/operator_requests.rs
//! Operator request endpoints.

use axum::{
    extract::{Path, Query, State},
    Json,
};
use serde::{Deserialize, Serialize};

use sera_db::operator_requests::OperatorRequestRepository;

use crate::error::AppError;
use crate::state::AppState;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OperatorRequestResponse {
    pub id: String,
    pub agent_id: String,
    pub agent_name: Option<String>,
    pub r#type: String,
    pub title: String,
    pub payload: serde_json::Value,
    pub status: String,
    pub response: Option<serde_json::Value>,
    pub created_at: String,
    pub resolved_at: Option<String>,
}

fn to_response(r: sera_db::operator_requests::OperatorRequestRow) -> OperatorRequestResponse {
    OperatorRequestResponse {
        id: r.id.to_string(),
        agent_id: r.agent_id,
        agent_name: r.agent_name,
        r#type: r.r#type,
        title: r.title,
        payload: r.payload,
        status: r.status,
        response: r.response,
        created_at: super::iso8601(r.created_at),
        resolved_at: r.resolved_at.map(super::iso8601),
    }
}

/// GET /api/operator-requests/pending/count
pub async fn pending_count(
    State(state): State<AppState>,
) -> Result<Json<serde_json::Value>, AppError> {
    let count = OperatorRequestRepository::count_pending(state.db.inner()).await?;
    Ok(Json(serde_json::json!({"count": count})))
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListRequestsQuery {
    pub status: Option<String>,
    pub agent_id: Option<String>,
    pub limit: Option<i64>,
}

/// GET /api/operator-requests
pub async fn list_requests(
    State(state): State<AppState>,
    Query(params): Query<ListRequestsQuery>,
) -> Result<Json<Vec<OperatorRequestResponse>>, AppError> {
    let limit = params.limit.unwrap_or(50).min(500);
    let rows = OperatorRequestRepository::list(
        state.db.inner(),
        params.status.as_deref(),
        params.agent_id.as_deref(),
        limit,
    )
    .await?;
    Ok(Json(rows.into_iter().map(to_response).collect()))
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RespondRequest {
    pub status: String,
    pub response: Option<serde_json::Value>,
}

/// POST /api/operator-requests/:id/respond
pub async fn respond_to_request(
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(body): Json<RespondRequest>,
) -> Result<Json<OperatorRequestResponse>, AppError> {
    let row = OperatorRequestRepository::respond(
        state.db.inner(),
        &id,
        &body.status,
        body.response.as_ref(),
    )
    .await?;
    Ok(Json(to_response(row)))
}

File: core/src/routes/operator-requests.ts
/**
 * Operator Requests Routes
 *
 * Bidirectional communication channel between SERA agents and operators (human or Claude Code).
 * Agents create requests when they need something they cannot do themselves.
 */

import { Router } from 'express';
import { pool } from '../lib/database.js';
import type { IntercomService } from '../intercom/IntercomService.js';
import { requireRole } from '../auth/authMiddleware.js';
import { rateLimitStub } from '../middleware/rateLimitStub.js';

export function createOperatorRequestsRouter(intercom?: IntercomService): Router {
  const router = Router();

  /**
   * GET /api/operator-requests/pending/count — Count pending requests (for badges)
   * NOTE: Registered before parameterised routes to avoid Express 5 shadowing.
   */
  router.get(
    '/pending/count',
    rateLimitStub,
    requireRole(['admin', 'operator']),
    async (_req, res) => {
      try {
        const { rows } = await pool.query(
          "SELECT COUNT(*)::int AS count FROM operator_requests WHERE status = 'pending'"
        );
        res.json({ count: rows[0]!.count });
      } catch (err) {
        res.status(500).json({ error: (err as Error).message });
      }
    }
  );

  /**
   * GET /api/operator-requests — List operator requests
   * Query params: status, agentId, limit
   */
  router.get('/', rateLimitStub, requireRole(['admin', 'operator']), async (req, res) => {
    try {
      const { status, agentId, limit: limitStr } = req.query;
      const limit = Math.min(Math.max(parseInt(String(limitStr || '50'), 10) || 50, 1), 200);

      const conditions: string[] = [];
      const params: unknown[] = [];

      if (status) {
        params.push(status);
        conditions.push(`status = $${params.length}`);
      }
      if (agentId) {
        params.push(agentId);
        conditions.push(`agent_id = $${params.length}`);
      }

      let query = 'SELECT * FROM operator_requests';
      if (conditions.length > 0) {
        query += ' WHERE ' + conditions.join(' AND ');
      }
      params.push(limit);
      query += ` ORDER BY created_at DESC LIMIT $${params.length}`;

      const { rows } = await pool.query(query, params);

      // Map snake_case to camelCase for frontend
      res.json(
        rows.map((r: Record<string, unknown>) => ({
          id: r.id,
          agentId: r.agent_id,
          agentName: r.agent_name,
          type: r.type,
          title: r.title,
          payload: r.payload,
          status: r.status,
          response: r.response,
          createdAt: r.created_at,
          resolvedAt: r.resolved_at,
        }))
      );
    } catch (err) {
      res.status(500).json({ error: (err as Error).message });
    }
  });

  /**
   * POST /api/operator-requests/:id/respond — Respond to a request
   * Body: { action: 'approved' | 'rejected' | 'resolved', response?: string | object }
   */
  router.post(
    '/:id/respond',
    rateLimitStub,
    requireRole(['admin', 'operator']),
    async (req, res) => {
      try {
        const { id } = req.params;
        const { action, response } = req.body as {
          action?: string;
          response?: unknown;
        };

        if (!action || !['approved', 'rejected', 'resolved'].includes(action)) {
          return res
            .status(400)
            .json({ error: 'action must be one of: approved, rejected, resolved' });
        }

        const responseJson =
          response != null
            ? typeof response === 'string'
              ? JSON.stringify({ message: response })
              : JSON.stringify(response)
            : null;

        const { rows, rowCount } = await pool.query(
          `UPDATE operator_requests
         SET status = $1, response = $2, resolved_at = NOW()
         WHERE id = $3 AND status = 'pending'
         RETURNING *`,
          [action, responseJson, id]
        );

        if (rowCount === 0) {
          return res.status(404).json({ error: 'Request not found or already resolved' });
        }

        const row = rows[0] as Record<string, unknown>;

        // Notify via Centrifugo so agents see the response in real-time
        if (intercom) {
          intercom
            .publishSystem('operator_request_response', {
              requestId: id,
              agentId: row.agent_id,
              action,
              response: responseJson ? JSON.parse(responseJson) : null,
              timestamp: new Date().toISOString(),
            })
            .catch(() => {});
        }

        res.json({
          id: row.id,
          agentId: row.agent_id,
          status: action,
          response: row.response,
          resolvedAt: row.resolved_at,
        });
      } catch (err) {
        res.status(500).json({ error: (err as Error).message });
      }
    }
  );

  return router;
}
